proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1097 @@
+  and less like pattern signatures.
+* `#420`_: Not yet accepted proposal that is an amendment to `#285`_, with a
+  very tiny, technical tweak to a definition.
+
+This proposal supersedes all of the proposals above, including the accepted
+ones. The goal of writing this is to provide a unified framework, so that
+we can design our language with cohesive treatment of scoped type variables,
+instead of simply an agglomeration of features.
+
+Motivation
+----------
+
+With GHC's powerful type-level programming features, we need powerful abilities
+to bring type variables into scope. The proposal defers to the individual proposals
+linked above for motivation for why we generally want these type-level features.
+Individual aspects of this unifying proposal are motivated near where they are
+introduced.
+
+Principles
+----------
+
+The design here is guided by several high-level principles. These principles are not
+sacro-sanct, such that they admit no exceptions. Instead, we strive to uphold these
+principles, but relax them when doing so is well motivated.
+
+.. _LSP:
+
+1. **Lexical Scoping Principle (LSP)**. For every occurrence of an identifier, it is possible to uniquely identify its binding site, without involving the type system. (From `#378`_.)
+
+   The LSP implies
+
+   .. _LSPC:
+
+   **Lexical Scoping Principle Corollary (LSPC)**. For every appearance of an identifier
+   in a Haskell program, it is possible to determine whether that appearance is a
+   binding site or an occurrence, without involving the type system.
+
+   Motivation: These principles mean that we can understand the binding
+   structure of a program without relying on type inference, important both for the
+   implementation of GHC and the sanity of programmers.
+
+   .. _LLSP:
+
+#. **Local Lexical Scoping Principle (LLSP)**. For every occurrence of an identifier, it is possible to determine
+   whether that appearance is a binding site or an occurrence, without looking to see what identifiers are
+   already in scope.
+
+   This is a variant of the LSPC_ that also prevents us from looking at the set of in-scope identifiers for determining
+   whether something is a binding site.
+
+   Motivation: Tracking the set of in-scope variables is laborious for human readers. (The compiler is already
+   doing this during name resolution.) This fact becomes even more poignant if we consider the possibility
+   of mixing the term-level and type-level namespaces (`#270`_) and need to think about clashes between type
+   variables and imported term variables.
+
+   .. _SUP:
+
+#. **Syntactic Unification Principle (SUP)**. In the absence of punning, there is no difference between type-syntax and term-syntax.
+   (From `#378`_.)
+
+   Motivation: The SUP keeps us forward-compatible with a possible future where the
+   distinction between term-syntax and type-syntax is removed.
+
+   .. _EVP:
+
+#. **Explicit Variable Principle (EVP)**. It is possible to write out all (specified)
```

Comment:
So many acronyms :slightly_frowning_face: . I think that the readability of this proposal would vastly improve if you expanded these acronyms wherever you use them.

Reply:
Indeed.